serverless/deploy/github/combine_witness_signatures/combine_witness_signatures.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// merge_witness_signatures is a tool to manage the checkpoint.witnessed file.
+//
+// Given the following files:
+// - checkpoint: a log checkpoint
+// - checkpoint.witnessed: a log checkpoint with additional witness signatures (optional)
+// - a collection of additional witnessed checkpoints
+// this tool will determine whether to update an existing checkpoint.witnessed
+// from the additional witnessed checkpoints, or replace it entirely by merging
+// witness cosignatures with the contents of the checkepoint file.
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/formats/checkpoints"
+	"golang.org/x/mod/sumdb/note"
+)
+
+var (
+	storageDir     = flag.String("storage_dir", "", "Root directory of the log.")
+	logKey         = flag.String("log_public_key", "", "Log's public key.")
+	witnessKeys    = flag.String("witness_public_key_files", "", "One or more space-separated globs matching all the known witness keys.")
+	numRequired    = flag.Uint("required_witness_sigs", 0, "Specifies the minimum number of valid witness signatures required to be present on the output to consider the merge successful.")
+	output         = flag.String("output", "", "Output file to write combined checkpoint to.")
+	deleteConsumed = flag.Bool("delete_consumed", true, "Whether to delete files containing merged checkpoints.")
+)
+
+func main() {
```

Comment:
The behaviour for promoting/merging checkpoints in here would benefit from some tests. There are a sufficient number of moving parts that even with two sets of eyes on it I couldn't rule out the possibility of bugs.

Reply:
yeah, I agree - my plan was do the broad strokes to get it working, and then factor funcs out and add tests